include/nbl/asset/IDescriptorSetLayout.h

```diff
@@ -58,310 +32,356 @@ enum E_DESCRIPTOR_TYPE : uint8_t
+				return Invalid;
+
+			const uint32_t foundIndex = found - m_bindingNumbers;
+			assert(foundIndex < m_count);
+			return foundIndex;
+		}
 
-		SBindingRedirect(const size_t _count) : count(_count)
+		inline binding_number_t getBindingNumber(const uint32_t index) const
 		{
-			bindings = std::make_unique<uint32_t[]>(count << 1);
-			offsets = bindings.get() + count;
+			assert(index < m_count);
+			return m_bindingNumbers[index];
 		}
 
-		// For a given binding number `binding` in `bindings`, return its offset in `offsets`.
-		inline uint32_t operator[](const uint32_t binding) const
+		inline core::bitflag<IShader::E_SHADER_STAGE> getStageFlags(const uint32_t index) const
 		{
-			const uint32_t index = searchForBinding(binding);
+			assert(index < m_count);
+			return m_stageFlags[index];
+		}
+
+		inline uint32_t getCount(const uint32_t index) const
+		{
+			assert(index < m_count);
+			return (index == 0u) ? m_storageOffsets[index].data : m_storageOffsets[index].data - m_storageOffsets[index - 1].data;
+		}
+
+		inline storage_offset_t getStorageOffset(const uint32_t index) const
+		{
+			assert(index < m_count);
+			return (index == 0u) ? 0u : m_storageOffsets[index - 1];
+		}
+
+		// The follwoing are merely convenience functions for one off use.
+		// If you already have an index (the result of `searchForBinding`) lying around use the above functions for quick lookups, and to avoid unnecessary binary searches.
+
+		inline core::bitflag<IShader::E_SHADER_STAGE> getStageFlags(const binding_number_t binding) const
+		{
+			const auto index = searchForBinding(binding);
 			if (index == Invalid)
 				return Invalid;
 
-			return offsets[index];
+			return getStageFlags(index);
 		}
 
-		// Returns index into the `bindings` and `offsets` array for the given binding number `binding.
-		// Assumes `bindings` is sorted.
-		inline uint32_t searchForBinding(const uint32_t binding) const
+		inline uint32_t getCount(const binding_number_t binding) const
 		{
-			if (!bindings || !offsets || (count == 0ull))
+			const auto index = searchForBinding(binding);
+			if (index == Invalid)
 				return Invalid;
 
-			auto found = std::lower_bound(bindings.get(), bindings.get() + count, binding);
+			return getDescriptorCount(index);
+		}
 
-			if ((found >= bindings.get() + count) || (*found != binding))
+		inline storage_offset_t getStorageOffset(const binding_number_t binding) const
+		{
+			const auto index = searchForBinding(binding);
+			if (index == Invalid)
 				return Invalid;
 
-			const uint32_t foundIndex = found - bindings.get();
-			assert(foundIndex < count);
-			return foundIndex;
+			return getStorageOffset(index);
 		}
 
-		std::unique_ptr<uint32_t[]> bindings;
-		uint32_t* offsets;
-		size_t count;
-	};
+		inline uint32_t getTotalCount() const { return (m_count == 0ull) ? 0u : m_storageOffsets[m_count - 1].data; }
 
-	public:
-		using sampler_type = SamplerType;
-
-		struct SBinding
+	private:
+		friend class IDescriptorSetLayout;
```

Comment:
why are you friending yourself? makes no sense...

Reply:
Declaring `IDescriptorSetLayout` a `friend` of `CBindingRedirect` so it has access to its `private`s like a friend does